src/ssh/cockpitsshrelay.c

```diff
@@ -896,96 +896,183 @@ do_password_auth (CockpitSshData *data)
+      g_autofree gchar *pub_key_path = NULL;
+      ssh_key priv_key = NULL;
+      ssh_key pub_key = NULL;
+
+      if (i == -1)
+        identity = g_strdup (libssh_identity);
+      else
         {
-          g_debug ("%s isn't accepted by the server", identity);
-          goto out;
+          identity = g_strdup_printf ("%s/.ssh/%s", g_get_home_dir (), default_identities[i]);
+          // No need to try the libssh identity twice, and we need to
+          // be precious with our tries because when we run into
+          // MaxAuthTries, libssh will hang.
+          if (g_strcmp0 (identity, libssh_identity) == 0)
+            continue;
         }
-    }
-  else if (rc == SSH_EOF)
-    {
-      g_debug ("Public key file %s doesn't exist or isn't readable", pub_key_path);
-    }
-  else
-    {
-      msg = ssh_get_error (data->session);
-      g_warning ("Error importing public key %s: %s", pub_key_path, msg);
-    }
 
-  rc = ssh_pki_import_privkey_file (identity, NULL, prompt_for_identity_password, data, &priv_key);
-  if (rc != SSH_OK)
-    {
-      g_debug ("Unable to import identity %s", identity);
-      goto out;
-    }
+      pub_key_path = g_strconcat (identity, ".pub", NULL);
+      rc = ssh_pki_import_pubkey_file (pub_key_path, &pub_key);
+      /* If the public key file exist and is readable, see if the identity is accepted by the server */
+      if (rc == SSH_OK)
+        {
+          rc = ssh_userauth_try_publickey (data->session, NULL, pub_key);
+          if (rc != SSH_AUTH_SUCCESS)
+            {
+              g_debug ("%s isn't accepted by the server", identity);
+              ssh_key_free (pub_key);
+              continue;
+            }
+        }
+      else if (rc == SSH_EOF)
+        {
+          g_debug ("Public key file %s doesn't exist or isn't readable", pub_key_path);
+        }
+      else
+        {
+          msg = ssh_get_error (data->session);
+          g_warning ("Error importing public key %s: %s", pub_key_path, msg);
+        }
 
-  rc = ssh_userauth_publickey (data->session, NULL, priv_key);
-  switch (rc)
-    {
-    case SSH_AUTH_SUCCESS:
-      g_debug ("%s: key auth succeeded", data->logname);
-      break;
-    case SSH_AUTH_DENIED:
-      g_debug ("%s: key auth failed", data->logname);
-      break;
-    case SSH_AUTH_PARTIAL:
-      g_message ("%s: key auth worked, but server wants more authentication",
-                 data->logname);
-      break;
-    case SSH_AUTH_AGAIN:
-      g_message ("%s: key auth failed: server asked for retry",
-                 data->logname);
-      break;
-    default:
-      msg = ssh_get_error (data->session);
-      g_message ("%s: couldn't key authenticate: %s", data->logname, msg);
-    }
+      struct CockpitSshPromptData pd = { data, identity, FALSE };
+      rc = ssh_pki_import_privkey_file (identity, NULL, prompt_for_identity_password, &pd, &priv_key);
+      if (rc == SSH_ERROR && pd.did_prompt)
+        rc = SSH_AUTH_DENIED;
+
+      if (rc == SSH_EOF)
```

Comment:
Hm can you make this an else if? It feels weird assigning the rc code and then checking against it further.

Reply:
Agreed, done.